src/v/kafka/server/group_manager.cc

```diff
@@ -1187,6 +1188,55 @@ group_manager::offset_fetch(offset_fetch_request&& r) {
     return group->handle_offset_fetch(std::move(r)).finally([group] {});
 }
 
+ss::future<offset_delete_response>
+group_manager::offset_delete(offset_delete_request&& r) {
+    auto error = validate_group_status(
+      r.ntp, r.data.group_id, offset_fetch_api::key);
+    if (error != error_code::none) {
+        co_return offset_delete_response(error);
+    }
+
+    auto group = get_group(r.data.group_id);
+    if (!group) {
+        co_return offset_delete_response(error_code::group_id_not_found);
+    }
+
+    std::vector<model::topic_partition> requested_deletions;
+    for (const auto& topic : r.data.topics) {
+        for (const auto& partition : topic.partitions) {
+            requested_deletions.emplace_back(
+              topic.name, partition.partition_index);
+        }
+    }
+
+    auto deleted_offsets = group->delete_offsets(requested_deletions);
+    co_await delete_offsets(group, deleted_offsets);
+
+    absl::flat_hash_set<model::topic_partition> deleted_offsets_set;
+    for (auto& tp : deleted_offsets) {
+        deleted_offsets_set.insert(std::move(tp));
+    }
+
+    absl::
+      flat_hash_map<model::topic, std::vector<offset_delete_response_partition>>
+        response_data;
+    for (const auto& tp : requested_deletions) {
+        auto error = kafka::error_code::none;
+        if (!deleted_offsets_set.contains(tp)) {
+            error = kafka::error_code::group_subscribed_to_topic;
```

Comment:
shouldn't this be done at the same time as calculating what to delete so that there is a consistent calculation without an intervening `co_await`?

Reply:
Its consistent because the only offsets not returned could only be missing from the set because of this condition. However I agree it is odd settings an error like this. Would be better maybe if the API returned offsets + error code or like you said calculate what to delete up front. I liked giving the responsibility of what to delete to the `group` class.